clientapi/routing/auth_fallback.go

```diff
@@ -0,0 +1,216 @@
+<script>
+if (window.onAuthDone) {
+    window.onAuthDone();
+} else if (window.opener && window.opener.postMessage) {
+    window.opener.postMessage("authDone", "*");
+}
+</script>
+</head>
+<body>
+    <div>
+        <p>Thank you!</p>
+        <p>You may now close this window and return to the application.</p>
+    </div>
+</body>
+</html>
+`
+
+// serveTemplate fills template data and serves it using http.ResponseWriter
+func serveTemplate(w http.ResponseWriter, templateHTML string, data map[string]string) {
+	t := template.Must(template.New("response").Parse(templateHTML))
+	if err := t.Execute(w, data); err != nil {
+		panic(err)
+	}
+}
+
+// AuthFallback implements GET on /auth/{authType}/fallback/web?session={sessionID}
+func AuthFallback(
+	w http.ResponseWriter, req *http.Request, authType string,
+	cfg config.Dendrite,
+) *util.JSONResponse {
+	sessionID := req.URL.Query().Get("session")
+
+	if sessionID == "" {
+		return writeErrorMessage(w, req,
+			"Session ID not provided",
+			http.StatusBadRequest,
+		)
+	}
+
+	serveRecaptcha := func() {
+		data := map[string]string{
+			"myUrl":   req.URL.String(),
+			"session": sessionID,
+			"siteKey": cfg.Matrix.RecaptchaPublicKey,
+		}
+		serveTemplate(w, recaptchaTemplate, data)
+	}
+
+	serveSuccess := func() {
+		data := map[string]string{}
+		serveTemplate(w, successTemplate, data)
+	}
+
+	if req.Method == http.MethodGet {
+		// Handle Recaptcha
+		if authType == authtypes.LoginTypeRecaptcha {
+			if err := checkRecaptchaEnabled(&cfg, w, req); err != nil {
+				return err
+			}
+
+			serveRecaptcha()
+			return nil
+		}
+		return &util.JSONResponse{
+			Code: http.StatusNotFound,
+			JSON: jsonerror.NotFound("Unknown auth stage type"),
+		}
+	} else if req.Method == http.MethodPost {
+		// Handle Recaptcha
+		if authType == authtypes.LoginTypeRecaptcha {
+			if err := checkRecaptchaEnabled(&cfg, w, req); err != nil {
+				return err
+			}
+
+			clientIP := req.RemoteAddr
+			err := req.ParseForm()
+			if err != nil {
+				res := httputil.LogThenError(req, err)
+				return &res
+			}
+
+			response := req.Form.Get("g-recaptcha-response")
+			if err := validateRecaptcha(&cfg, response, clientIP); err != nil {
+				util.GetLogger(req.Context()).Error(err)
+				return err
+			}
+
+			// Success. Add recaptcha as a completed login flow
+			AddCompletedSessionStage(sessionID, authtypes.LoginTypeRecaptcha)
+
+			serveSuccess()
+			return nil
+		}
+
+		return &util.JSONResponse{
+			Code: http.StatusNotFound,
+			JSON: jsonerror.NotFound("Unknown auth stage type"),
+		}
+	}
+	return &util.JSONResponse{
+		Code: http.StatusMethodNotAllowed,
+		JSON: jsonerror.NotFound("Bad method"),
+	}
+}
+
+// checkRecaptchaEnabled creates an error response if recaptcha is not usable on homeserver.
+func checkRecaptchaEnabled(
+	cfg *config.Dendrite,
+	w http.ResponseWriter,
+	req *http.Request,
+) *util.JSONResponse {
+	if cfg.Matrix.RecaptchaEnabled {
+		if cfg.Matrix.RecaptchaPublicKey == "" {
```

Comment:
Perhaps we don't need this `RecaptchaPublicKey` check?

Reply:
It would be better to keep it. It can cause failure for client and it is good to know what happened. It's a good edge case to handle.

Reply:
Hmmm, then do you also check `RecaptchaPrivateKey`?

Reply:
Not currently, but we should add that. I didn't know there is a PrivateKey too. We can check both here.

Reply:
I think checks for `RecaptchaPublicKey` and `RecaptchaPrivateKey` (and things alike) should be done during startup if `RecaptchaEnabled` is set to `true`, so other places can be sure about this.

Reply:
Yes, this would be better. I'll make a seperate PR to add these checks to startup scripts.
I'll remove from here.